File: programming/problem_7/problem_7.go
package main

import (
	"fmt"
)

func main() {
	numbers := []int{1, 2, 3, 4, 5}
	evenAverage, oddAverage := calculateEvenAndOddAverages(numbers)
	fmt.Printf("oddAverage: %f\n", oddAverage)
	fmt.Printf("evenAverage: %f\n", evenAverage)
}

func calculateEvenAndOddAverages(numbers []int) (float64, float64) {
	evenNumbers := filter(numbers, func(n int) bool { return n%2 == 0 })
	oddNumbers := filter(numbers, func(n int) bool { return n%2 != 0 })

	return calculateAverage(evenNumbers), calculateAverage(oddNumbers)
}

func filter(numbers []int, condition func(int) bool) []int {
	var filtered []int
	for _, num := range numbers {
		if condition(num) {
			filtered = append(filtered, num)
		}
	}
	return filtered
}		

func calculateAverage(numbers []int) float64 {
	if len(numbers) == 0 {
		return 0.0
	}

	sum := 0
	for _, num := range numbers {
		sum += num
	}
	return float64(sum) / float64(len(numbers))
}

File: programming/problem_7/problem_7.rb
def main
    numbers = [1, 2, 3, 4, 5]
    even_average, odd_average = calculate_even_and_odd_averages(numbers)
    puts "oddAverage: #{odd_average}"
    puts "evenAverage: #{even_average}"
end
  
def calculate_even_and_odd_averages(numbers)
    [
        calculate_average(numbers.select { |n| n.even? }),
        calculate_average(numbers.select { |n| n.odd? })
    ]
end
  
def calculate_average(numbers)
    return 0.0 if numbers.empty?

    numbers.sum.to_f / numbers.length
end
  
main
  